rust/plugin-lib/src/state_cache.rs

```diff
@@ -0,0 +1,517 @@
+        let ctx = PluginCtx {
+            state: &mut self.state,
+            peer: peer,
+        };
+        self.handler.idle(ctx, token);
+    }
+}
+
+pub fn mainloop<H: Handler>(handler: &mut H) {
+    let mut my_handler = MyHandler {
+        handler: handler,
+        state: MyState::default(),
+    };
+    plugin_base::mainloop(&mut my_handler);
+}
+
+impl<'a, S: Default + Clone> PluginCtx<'a, S> {
+
+    pub fn get_path(&self) -> Option<&PathBuf> {
+        match self.state.path {
+            Some(ref p) => Some(p),
+            None => None,
+        }
+    }
+
+    pub fn add_scopes(&self, scopes: &Vec<Vec<String>>) {
+        self.peer.add_scopes(&self.state.view_id, scopes)
+    }
+
+    pub fn update_spans(&self, start: usize, len: usize, spans: &[ScopeSpan]) {
+        self.peer.update_spans(&self.state.view_id, start, len, self.state.rev, spans)
+    }
+
+    /// Determines whether an incoming request (or notification) is pending. This
+    /// is intended to reduce latency for bulk operations done in the background.
+    pub fn request_is_pending(&self) -> bool {
+        self.peer.request_is_pending()
+    }
+
+    /// Schedule the idle handler to be run when there are no requests pending.
+    pub fn schedule_idle(&mut self, token: usize) {
+        self.peer.schedule_idle(token);
+    }
+
+    /// Find an entry in the cache by line num. On return `Ok(i)` means entry
+    /// at index `i` is an exact match, while `Err(i)` means the entry would be
+    /// inserted at `i`.
+    fn find_line(&self, line_num: usize) -> Result<usize, usize> {
+        self.state.state_cache.binary_search_by(|probe| probe.line_num.cmp(&line_num))
+    }
+
+    /// Find an entry in the cache by offset. Similar to `find_line`.
+    fn find_offset(&self, offset: usize) -> Result<usize, usize> {
+        self.state.state_cache.binary_search_by(|probe| probe.offset.cmp(&offset))
+    }
+
+    /// Get state at or before given line number. Returns line number, offset,
+    /// and user state.
+    pub fn get_prev(&self, line_num: usize) -> (usize, usize, S) {
+        if line_num > 0 {
+            let mut ix = match self.find_line(line_num) {
+                Ok(ix) => ix,
+                Err(0) => return (0, 0, S::default()),
+                Err(ix) => ix - 1,
+            };
+            loop {
+                let item = &self.state.state_cache[ix];
+                if let Some(ref s) = item.user_state {
+                    return (item.line_num, item.offset, s.clone());
+                }
+                if ix == 0 { break; }
+                ix -= 1;
+            }
+        }
+        (0, 0, S::default())
+    }
+
+    /// Get the state at the given line number, if it exists in the cache.
+    pub fn get(&self, line_num: usize) -> Option<&S> {
+        if let Ok(ix) = self.find_line(line_num) {
+            self.state.state_cache[ix].user_state.as_ref()
+        } else {
+            None
+        }
```

Comment:
golf:
```rust
self.find_line(line_num).ok()
    .and_then(|ix| self.state.state_cache[ix].user_state.as_ref())
```

Reply:
Done. Here I agree, but often I find the written-out form more readable than the combinators.